### PR TITLE
Use docker-compose on travis for linux builds

### PR DESCRIPTION
**Pull request by andreas-bok-sociomantic**
The PR changes the existing Travis CI configuration and 
scripts to use docker-compose for all linux based builds.

The main reason is the ability to build and deploy packages 
for Ubuntu Xenial, which is not possible in the native 
travis build environment.

### CODE REVIEW

File: tests/unit/connect_uds.c [hunk 1]
```diff
@@ -49,6 +49,8 @@ int main(int argc, char *argv[])
   (void)argc;
   (void)argv;
 
+  SKIP_IF(getenv("MYSQL_DOCKER_CONTAINER") != NULL);
```
Comment (leandro-lucarella-sociomantic) [inline tests/unit/connect_uds.c]: It fails how? Can it be fixed? I wouldn't disable tests unless there is no other way.

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+
+services:
+  db:
+    image: mysql:5.7
+    restart: always
+    environment:
+      MYSQL_ROOT_PASSWORD: pass
+      MYSQL_DATABASE: test
+      MYSQL_USER: user
+      MYSQL_PASSWORD: pass
```
Comment (leandro-lucarella-sociomantic) [inline docker-compose.yml]: Why this doesn't use the `.env` file? DRY.

File: docker/.env [hunk 3]
```diff
@@ -0,0 +1,11 @@
+# Docker container environment variables
+# The file is used for the docker-compose.yml::env_file setting
```
Comment (leandro-lucarella-sociomantic) [inline docker/.env]: BTW, why `.env` and not just `env` or `docker-compose.env`? Why hidden?

File: docker/docker-compose.ubuntu.xenial.yml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# Docker compose file to override configuration for centos based images
```
Comment (leandro-lucarella-sociomantic) [inline docker/docker-compose.ubuntu.xenial.yml:1]: Is this a standard override mechanism or how is this file used?

File: docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
+# Configuration file to run ci tests in docker containers
```
Comment (leandro-lucarella-sociomantic) [inline docker-compose.yml:1]: I would include the changes needed to use these files as part of the same commit. It is really quite complicated to figure out the interaction between these files or how will they be used in this commit.

File: docker/Dockerfile.centos.7 [hunk 6]
```diff
@@ -1,34 +1,12 @@
-# Docker build file to create an image running centos7 with MySQL and build
-# tools installed
-
-FROM centos/mysql-56-centos7:latest
+# Docker build file to create an image with centos 7
+FROM centos:7
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.centos.7:2]: Same as previous commit, I can't see how these new files are used. I guess they aren't so far, which is weird.

File: docker/docker-compose.centos.7.yml [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# Docker compose file to override configuration for centos based images
```
Comment (leandro-lucarella-sociomantic) [inline docker/docker-compose.centos.7.yml:1]: Same question about this override file.

File: .travis.yml [hunk 8]
```diff
@@ -4,20 +4,14 @@ language: cpp
 cache: ccache
 
 ## Enable sudo
-sudo: required
+sudo: false
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml]: Whot? Why isn't sudo needed when using docker compose? According to https://docs.travis-ci.com/user/docker/ you need `sudo: required` to run docker.

File: .travis.yml [hunk 9]
```diff
@@ -96,12 +87,18 @@ before_script:
...
+after_script:
+  - ./travis_configure.sh after_script
+
+## Deploy debian packages to bintray
+##
+## deb package built on ubuntu xenial with gcc
 deploy:
     provider: script
-    script: beaver bintray upload -N pkg/deb/*.deb
+    script: beaver bintray upload -o -N pkg/deb/*.deb
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml]: Is this really related to this commit?

File: .travis.yml [hunk 10]
```diff
@@ -96,12 +87,18 @@ before_script:
...
     skip_cleanup: true
     on:
-        tags: true
-        repo: sociomantic-tsunami/libdrizzle-redux
-        condition: $TRAVIS_OS_NAME = 'linux' && $CC = 'gcc' && $DIST_PACKAGE_TARGET = 'DEB'
+        tags: true # must be a git tag
+        repo: sociomantic-tsunami/libdrizzle-redux # must be upstream
+        condition: $TRAVIS_OS_NAME.$DIST_NAME.$DIST_VERSION = 'linux.ubuntu.xenial' && $CC = 'gcc' && $MAKE_TARGET =~ 'deb' && $DEPLOY_PACKAGE == 'true'
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml]: Why concatenating `$TRAVIS_OS_NAME.$DIST_NAME.$DIST_VERSION` but not `$CC`, `$MAKE_TARGET` and `$DEPLOY_PACKAGE`?

File: travis_configure.sh [hunk 11]
```diff
@@ -15,27 +15,30 @@ print_error_msg ()
...
 #   - sed, (libtool), openssl, mysql
 # Returns 0
 before_install()
 {
     if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
-        if [[ "$DIST_PACKAGE_TARGET" == "DEB" ]]; then
-            gem install fpm -v 1.8.1
+        # build images
+        docker-compose -f ./docker-compose.yml \
+            -f ./docker/docker-compose.$DIST_NAME.$DIST_VERSION.yml \
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh:37]: OK, I see now, nice trick. I think the 3 last commits (including this one) should be squashed into one.

File: travis_configure.sh [hunk 12]
```diff
@@ -15,27 +15,30 @@ print_error_msg ()
...
+            build
 
+        # jfrog dependency
+        if [[ "$MAKE_TARGET" =~ "deb" ]]; then
             if [[ -n "$TRAVIS_TAG" && $TRAVIS_REPO_SLUG == "sociomantic-tsunami/libdrizzle-redux" ]]; then
                 curl -XGET -L -k 'https://api.bintray.com/content/jfrog/jfrog-cli-go/$latest/jfrog-cli-linux-amd64/jfrog?bt_package=jfrog-cli-linux-amd64' > /tmp/jfrog ;
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh:43]: Maybe you want to use `sociomantictsunami/base` image now (it comes with `jfrog`)? Or was there a problem with that?

File: travis_configure.sh [hunk 13]
```diff
@@ -148,13 +131,26 @@ run_tests()
...
 }
 
+# Invoked by the travis hook script
+#
+# - if the is run in docker stop containers and remove containers, networks,
+#   volumes, and images created by `up`.
+after_script()
+{
+    if [[ "$TRAVIS_OS_NAME" == "linux" ]]; then
+        docker-compose down
```
Comment (leandro-lucarella-sociomantic) [inline travis_configure.sh]: Even when this might be nice, I guess it's not really necessary (at least for travis), the VM will be killed anyway.

**Review by leandro-lucarella-sociomantic — COMMENTED** [review]
Generally speaking, ME LIKES where this is going!

File: tests/unit/connect_uds.c (hunk 1 above)
Comment (andreas-bok-sociomantic) [reply tests/unit/connect_uds.c]: It is a workaround, since I couldn't get the connection to the db container working using a socket. 
But true it should of course be fixed before it is merged.

File: .travis.yml (hunk 8 above)
Comment (andreas-bok-sociomantic) [reply .travis.yml]: I see. Strangely it builds with `sudo: false`. With the example they use `sudo` is required but AFAICS only because of the `sudo apt` commands. The only sudo command which runs in the native environment is the [`sudo cp /tmp/jfrog /usr/local/bin/jfrog ;` ](https://github.com/andreas-bok-sociomantic/libdrizzle-redux/blob/acdd40467a897e1a3d0eb144a9ac35d783ca7d81/travis_configure.sh#L45)
The sudo command in [`enable_mysqlbinlog()`]() is run in the OSX virtual environment but by the CI service MacStadium, https://www.macstadium.com/blog/how-travis-ci-halved-boot-times-for-os-x-and-ios-testing-at-macstadium/
Furthermore having `sudo: false` makes the build start in a container. https://docs.travis-ci.com/user/migrating-from-legacy/
However the bottleneck is still the OSX build so overall it doesn't make the whole build finish faster

File: .travis.yml (hunk 9 above)
Comment (andreas-bok-sociomantic) [reply .travis.yml]: Nope. Will remove, but quite convenient for testing since otherwise you would to change the git `tag` each time you wanna test the deployment

File: .travis.yml (hunk 10 above)
Comment (andreas-bok-sociomantic) [reply .travis.yml]: You mean making one long string to compare with? e.g. 

    $TRAVIS_OS_NAME.$DIST_NAME.$DIST_VERSION.$CC.$MAKE_TARGET.$DEPLOY_PACKAGE =~ 'linux.ubuntu.xenial.gcc.deb.true'

File: travis_configure.sh (hunk 12 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh:43]: I guess I could, but all the **D** libs and packages which are installed would be obsolete for this case. 
Also currently the deployment step is handled by `travis` as defined in the `.travis.yml` file. So not much seems to be gained by using `sociomantictsunami/base`. 
However we could consider adding these Dockerfiles to the `sociomantictsunami` repo at `hub.docker.com`

File: travis_configure.sh (hunk 13 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh]: True, I will try to test if it all the same without `docker-compose down`

File: docker-compose.yml (hunk 2 above)
Comment (andreas-bok-sociomantic) [reply docker-compose.yml]: It should. Some env variables are not relevant for the db container. Multiple files can be specified for `env_file` so will separate the vars into shared and container specific.

File: docker/.env (hunk 3 above)
Comment (andreas-bok-sociomantic) [reply docker/.env]: Haha, just a consequence of following the example at docker.com blindly.

File: docker/docker-compose.ubuntu.xenial.yml (hunk 4 above)
Comment (andreas-bok-sociomantic) [reply docker/docker-compose.ubuntu.xenial.yml:1]: Is this a standard override mechanism or how is this file used?

Yes, cf. https://docs.docker.com/compose/extends/#example-use-case

But guess the files could be generated dynamically given that they only specify which Dockerfile to use

File: travis_configure.sh (hunk 11 above)
Comment (andreas-bok-sociomantic) [reply travis_configure.sh:37]: Sure, guess I was more focused on the extent than the logics of the changes.

File: .travis.yml (hunk 8 above)
Comment (leandro-lucarella-sociomantic) [reply .travis.yml]: Have you seen the link I posted? :)

I think this is it:

> Travis CI automatically routes builds to run on Trusty sudo: required when services: docker is configured. We do not currently support use of Docker on OS X.

I wonder why just before that note they say you should use `sudo: required` if in the next line they say is automatically done when `services: - docker` is used :thinking:

Comment (leandro-lucarella-sociomantic) [reply .travis.yml]: Yup, it looks like it's a sudo: https://travis-ci.org/sociomantic-tsunami/libdrizzle-redux/jobs/276921313#L125

File: .travis.yml (hunk 10 above)
Comment (leandro-lucarella-sociomantic) [reply .travis.yml]: Yes (with `=`, not `=~`). Actually I think is safer to compare every string individually, as if you have weird combinations the concatenation could match when the individual parts wont. I think is unlikely enough to be acceptable, but I really don't see any reason not to compare every variable individually instead.

File: travis_configure.sh (hunk 12 above)
Comment (leandro-lucarella-sociomantic) [reply travis_configure.sh:43]: The `base` image doesn't have any D stuff :)

Comment (andreas-bok-sociomantic) [reply travis_configure.sh:43]: Ah, got things confused a bit after looking at cachalot 😀

**Comment by andreas-bok-sociomantic** [comment]
Updated based on comments from @leandro-lucarella-sociomantic

File: .travis.yml [hunk 14]
```diff
@@ -3,21 +3,16 @@ language: cpp
 ## Enable cache for dependencies and directories
 cache: ccache
 
-## Enable sudo
-sudo: required
+## Enable sudo (set explicitly although environment's with docker is sudo
+#  enabled  by default
+sudo: true
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml]: Why not `required` as it was? :)

File: docker/Dockerfile.ubuntu.xenial [hunk 15]
```diff
@@ -0,0 +1,35 @@
+# Dockerfile for building an image with ubuntu xenial
+FROM sociomantictsunami/base
+MAINTAINER andreas-bok-sociomantic <andreas.bok@sociomantic.com>
+
+WORKDIR /root
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.ubuntu.xenial]: :scream:

File: docker/Dockerfile.ubuntu.xenial [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+# install apt-fast to download required packages in parallel
+RUN echo "deb http://ppa.launchpad.net/saiarcot895/myppa/ubuntu xenial main" >> /etc/apt/sources.list
+RUN apt-key adv --keyserver keyserver.ubuntu.com --recv-keys DC058F40
+RUN apt-get update
+RUN apt-get install --no-install-recommends -y apt-fast
+
+# install required packages
+RUN apt-fast -y install --no-install-recommends \
+    libssl-dev autoconf libtool make automake zlib1g-dev
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.ubuntu.xenial]: Every `RUN` command in Docker creates a new layer, it is usually recommended to create some script and just use one `RUN` (or if it's just a few commands use `RUN cmd && cmd && cmd`). Not a big deal, just FYI.

**Review by leandro-lucarella-sociomantic — COMMENTED** [review]
Just a few more comments, but changes look good in general

File: .travis.yml (hunk 14 above)
Comment (andreas-bok-sociomantic) [reply .travis.yml]: Ah sure, should be `required`

File: docker/Dockerfile.ubuntu.xenial (hunk 15 above)
Comment (andreas-bok-sociomantic) [reply docker/Dockerfile.ubuntu.xenial]: hehe, will change to a less contested dir

File: docker/Dockerfile.ubuntu.xenial (hunk 16 above)
Comment (andreas-bok-sociomantic) [reply docker/Dockerfile.ubuntu.xenial]: Oki, you mean a new temp image will be created for each RUN cmd? 

I will wrap the cmds  in a script

File: docker/Dockerfile.ubuntu.xenial (hunk 15 above)
Comment (leandro-lucarella-sociomantic) [reply docker/Dockerfile.ubuntu.xenial]: Afterwards you change it to `libdrizzle-redux` or something. I wonder why you set it at all.

File: docker/Dockerfile.ubuntu.xenial (hunk 16 above)
Comment (leandro-lucarella-sociomantic) [reply docker/Dockerfile.ubuntu.xenial]: Yup

**Comment by andreas-bok-sociomantic** [comment]
Updated based on comments from @leandro-lucarella-sociomantic 
Additional changes:
- moved the contents of `install_dependencies.sh` to a function in `container_entrypoint.sh`
- removed obsolete and out-commented install of **fpm** since it is already available in `sociomantictsunami/base`

File: docker/Dockerfile.ubuntu.xenial [hunk 17]
```diff
@@ -0,0 +1,10 @@
+# Dockerfile for building an image with ubuntu xenial
+FROM sociomantictsunami/base
+MAINTAINER andreas-bok-sociomantic <andreas.bok@sociomantic.com>
+
+WORKDIR /home
```
Comment (leandro-lucarella-sociomantic) [inline docker/Dockerfile.ubuntu.xenial]: I wonder why you need to set a workdir at all.

Comment (andreas-bok-sociomantic) [reply docker/Dockerfile.ubuntu.xenial]: Hmm, isn't `root` the default user in the images, which means that `$HOME` is  `/root`?

Comment (andreas-bok-sociomantic) [reply docker/Dockerfile.ubuntu.xenial]: Cf, https://docs.docker.com/engine/reference/run/#user and https://docs.docker.com/engine/reference/run/#workdir

**Comment by andreas-bok-sociomantic** [comment]
Updated
- fixed issue with socket connection by adding a volume in which the mysql socket is shared between the containers

File: docker/Dockerfile.ubuntu.xenial (hunk 17 above)
Comment (leandro-lucarella-sociomantic) [reply docker/Dockerfile.ubuntu.xenial]: When building the image it is root. But the question is why you need to know which is the work directory? Are you writing any files to it?

I think you are mixing up stuff, this is image creating, for running you just pass `-u`. The `USER` and `WORKDIR` commands make (more?) sense when using docker to actually deploy stuff, but for testing you should just run `docker run -u user -w workdir` if you want to set it to anything in particular, you don't have to do anything at the image creation unless you need it to actually create the image.

Comment (andreas-bok-sociomantic) [reply docker/Dockerfile.ubuntu.xenial]: > Are you writing any files to it?

Only writing the init script `ubuntu_xenial_image_run.sh` to `/usr/local/bin/`

> if you want to set it to anything in particular, you don't have to do anything at the image creation unless you need it to actually create the image.

Ah, I can set the working directory in the `docker-compose.yml` file, https://docs.docker.com/compose/compose-file/#domainname-hostname-ipc-mac_address-privileged-read_only-shm_size-stdin_open-tty-user-working_dir

**Comment by andreas-bok-sociomantic** [comment]
Updated
- removed `WORK_DIR` steps from Dockerfiles
- specify the working directory in `docker-compose.yml` instead

**Comment by leandro-lucarella-sociomantic** [comment]
Is there any ETA to finalize this PR?

**Comment by leandro-lucarella-sociomantic** [comment]
Also, can you add a "Fixes #239" in the commit that removes building for trusty, so it automatically gets closed when this is merged?

File: .travis.yml [hunk 18]
```diff
@@ -31,58 +25,55 @@ services:
...
-#
-#  To speed up installation of brew formulas on osx auto update for homebrew is
-#  turned off by setting HOMEBREW_NO_AUTO_UPDATE=1
 env:
   global:
-    - MYSQL_SERVER=127.0.0.1
+    - COMPILER_VERSION=
+    - DEPLOY_PACKAGE=false
+    - DIST_NAME=ubuntu
+    - DIST_VERSION=xenial
```
Comment (leandro-lucarella-sociomantic) [inline .travis.yml:33]: I'd mention that this commit also stops building for trusty and starts building for xenial.

**Comment by andreas-bok-sociomantic** [comment]
> Is there any ETA to finalize this PR?

I will address the comments, and then if everything is ok, it will be part of the [`v5.7.1`](https://github.com/sociomantic-tsunami/libdrizzle-redux/milestone/34) release

**Comment by andreas-bok-sociomantic** [comment]
@leandro-lucarella-sociomantic addressed your comments. If all is good then let's merge! :)

**Review by leandro-lucarella-sociomantic — APPROVED** [review]
